File: sdk/typescript/src/clients/reviews.ts
/**
 * Reviews resource client.
 *
 * Design Reference: DR-5
 * Requirements: 9.3
 */

import type { HTTPTransport } from '../transport.js';
import type { Review } from '../types/index.js';

/**
 * Parse ISO date string to Date object.
 */
function parseDate(dateStr: string): Date {
  return new Date(dateStr.replace(/Z$/, '+00:00'));
}

/**
 * Get value from object, trying camelCase first then snake_case.
 */
function getValue<T>(
  data: Record<string, unknown>,
  camel: string,
  snake: string,
  defaultValue?: T
): T {
  if (camel in data) {
    return data[camel] as T;
  }
  return (data[snake] ?? defaultValue) as T;
}

/**
 * Client for pull request review operations.
 *
 * Design Reference: DR-5
 */
export class ReviewsClient {
  private transport: HTTPTransport;

  /**
   * Initialize the reviews client.
   *
   * @param transport - HTTP transport for making requests
   */
  constructor(transport: HTTPTransport) {
    this.transport = transport;
  }

  /**
   * Submit a review for a pull request.
   *
   * The PR author cannot approve their own PR.
   *
   * @param repoId - The repository identifier
   * @param prId - The pull request identifier
   * @param verdict - "approve", "request_changes", or "comment"
   * @param body - Optional review comment
   * @returns Review object with reviewId
   * @throws AuthenticationError if signature is invalid
   * @throws ValidationError if self-approval attempted
   * @throws NotFoundError if pull request not found
   *
   * Requirements: 9.3
   */
  async create(
    repoId: string,
    prId: string,
    verdict: 'approve' | 'request_changes' | 'comment',
    body?: string | null
  ): Promise<Review> {
    const requestBody: Record<string, unknown> = {
      repoId,
      prId,
      verdict,
      body: body ?? null,
    };

    const response = await this.transport.signedRequest(
      'POST',
      `/v1/repos/${repoId}/pulls/${prId}/reviews`,
      'pr_review',
      requestBody
    );

    const data = (response.data ?? {}) as Record<string, unknown>;
    return {
      reviewId: data.reviewId as string,
      prId: data.prId as string,
      reviewerId: data.reviewerId as string,
      verdict: data.verdict as 'approve' | 'request_changes' | 'comment',
      body: (data.body as string | null) ?? null,
      createdAt: parseDate(data.createdAt as string),
    };
  }

  /**
   * List reviews for a pull request.
   *
   * @param repoId - The repository identifier
   * @param prId - The pull request identifier
   * @returns List of Review objects
   * @throws NotFoundError if pull request not found
   *
   * Requirements: 9.3
   */
  async list(repoId: string, prId: string): Promise<Review[]> {
    const response = await this.transport.unsignedRequest(
      'GET',
      `/v1/repos/${repoId}/pulls/${prId}/reviews`
    );

    const data = (response.data ?? {}) as Record<string, unknown>;
    // Handle both list and dict responses
    let reviews: Record<string, unknown>[];
    if (Array.isArray(data)) {
      reviews = data as Record<string, unknown>[];
    } else {
      reviews = (data.reviews ?? []) as Record<string, unknown>[];
    }

    return reviews.map((review) => ({
      reviewId: getValue<string>(review, 'reviewId', 'review_id', ''),
      prId: getValue<string>(review, 'prId', 'pr_id', ''),
      reviewerId: getValue<string>(review, 'reviewerId', 'reviewer_id', ''),
      verdict: review.verdict as 'approve' | 'request_changes' | 'comment',
      body: (review.body as string | null) ?? null,
      createdAt: parseDate(getValue<string>(review, 'createdAt', 'created_at', '')),
    }));
  }
}

File: sdk/rust/src/clients/reviews.rs
//! Reviews resource client.
//!
//! Design Reference: DR-5
//! Requirements: 9.3

use std::collections::HashMap;
use std::sync::Arc;

use serde_json::Value;

use crate::error::Error;
use crate::transport::HttpTransport;
use crate::types::Review;

/// Client for pull request review operations.
pub struct ReviewsClient {
    transport: Arc<HttpTransport>,
}

impl ReviewsClient {
    /// Create a new reviews client.
    pub fn new(transport: Arc<HttpTransport>) -> Self {
        Self { transport }
    }

    /// Submit a review on a pull request.
    ///
    /// # Arguments
    ///
    /// * `repo_id` - The repository identifier
    /// * `pr_id` - The pull request identifier
    /// * `verdict` - Review verdict: "approve", "request_changes", or "comment"
    /// * `body` - Optional review comment
    ///
    /// # Returns
    ///
    /// Review object
    ///
    /// # Errors
    ///
    /// Returns an error if the review fails.
    ///
    /// Requirements: 9.3
    pub async fn create(
        &self,
        repo_id: &str,
        pr_id: &str,
        verdict: &str,
        body: Option<&str>,
    ) -> Result<Review, Error> {
        let mut request_body: HashMap<String, Value> = HashMap::new();
        request_body.insert("repoId".to_string(), Value::String(repo_id.to_string()));
        request_body.insert("prId".to_string(), Value::String(pr_id.to_string()));
        request_body.insert("verdict".to_string(), Value::String(verdict.to_string()));
        request_body.insert(
            "body".to_string(),
            body.map_or(Value::Null, |b| Value::String(b.to_string())),
        );

        let response: Value = self
            .transport
            .signed_request(
                "POST",
                &format!("/v1/repos/{repo_id}/pulls/{pr_id}/reviews"),
                "review_create",
                request_body,
            )
            .await?;

        let data = response
            .get("data")
            .ok_or_else(|| Error::Http("Missing data in response".to_string()))?;

        serde_json::from_value(data.clone()).map_err(Error::from)
    }

    /// List reviews for a pull request.
    ///
    /// # Arguments
    ///
    /// * `repo_id` - The repository identifier
    /// * `pr_id` - The pull request identifier
    ///
    /// # Returns
    ///
    /// List of Review objects
    ///
    /// # Errors
    ///
    /// Returns an error if the request fails.
    pub async fn list(&self, repo_id: &str, pr_id: &str) -> Result<Vec<Review>, Error> {
        let response: Value = self
            .transport
            .unsigned_request::<Value>(
                "GET",
                &format!("/v1/repos/{repo_id}/pulls/{pr_id}/reviews"),
                None,
                None::<&()>,
            )
            .await?;

        let data = response
            .get("data")
            .ok_or_else(|| Error::Http("Missing data in response".to_string()))?;

        let reviews = data
            .get("reviews")
            .ok_or_else(|| Error::Http("Missing reviews in response".to_string()))?;

        serde_json::from_value(reviews.clone()).map_err(Error::from)
    }
}
